backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/Linker.kt

```diff
@@ -170,8 +175,36 @@ internal class Linker(val context: Context) {
         return executable
     }
 
+    private fun determineLinkerInput(objectFiles: List<ObjectFile>): LinkerInput {
+        val caches = determineCachesToLink(context)
+        // Since we have several linker stages that involve caching,
+        // we should detect cache usage early to report errors correctly.
+        val cachingInvolved = caches.static.isNotEmpty() || caches.dynamic.isNotEmpty()
+        val config = context.config
+        return when {
+            config.produce == CompilerOutputKind.STATIC_CACHE -> {
+                // Do not link static cache dependencies.
+                LinkerInput(objectFiles, CachesToLink(emptyList(), caches.dynamic), emptyList(), cachingInvolved)
+            }
+            config.cacheSupport.disableCachesPreLink || caches.static.isEmpty() -> {
+                LinkerInput(objectFiles, caches, emptyList(), cachingInvolved)
+            }
+            else -> {
```

Comment:
We don't actually need prelinking caches when producing anything but static libraries/frameworks.